New Package - Phase2 L1CaloTrigger

rekovic opened the pull request:
#### PR description:  New package for Phase2 L1Trigger 

Contains only Standalone EG in Barrel and EndCap.  These STA objects are needed for the L1T  Correlator and ParticleFlow, which will be PRed afterwards.

These currently hold 3 data files, which will be moved to Externals during the code review.

rekovic committed:
Phase2 L1CaloTrigger - only Standalone EG in Barrel and EndCap.

Co-authored-by: Gianluca <gianluca.cerminara@cern.ch>
Co-authored-by: Cecile Caillol <Cecile.Caillol@cern.ch>

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented:
will move the data files.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/15489

- This PR adds an extra 328KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/15489/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/15489/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
Code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/15492

- This PR adds an extra 328KB to repository

cmsbuild commented:
A new Pull Request was created by  @rekovic  for master.

It involves the following packages:

L1Trigger/L1CaloTrigger


The following packages do not have a category, yet:

L1Trigger/L1CaloTrigger
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@cmsbuild can you please review it and eventually sign? Thanks.
@Martin-Grunewald this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

silviodonato commented:
assign l1

silviodonato commented:
assign upgrade

cmsbuild commented:
New categories assigned: upgrade,l1

@benkrikler,@rekovic,@kpedro88 you have been requested to review this Pull request/Issue and eventually sign? Thanks

kpedro88 commented:
@rekovic in addition to moving the data files, the commits in this PR should be rebased or squashed so that the data files never enter the cmssw repository history

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+}
+
+bool L1EGCrystalClusterEmulatorProducer::passes_looseTkss(float pt, float ss) {
+  if ((e0_looseTkss - e1_looseTkss * TMath::Exp(-e2_looseTkss * pt)) > ss)
```
and simplify

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+#include "DataFormats/Math/interface/deltaPhi.h"
+
+// we sort the clusters in pt
+bool compare_cluster_pt(const l1t::HGCalMulticluster *cl1, const l1t::HGCalMulticluster *cl2);
```
no need for forward declaration 2 lines above the actual declaration

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+// we sort the clusters in pt
+bool compare_cluster_pt(const l1t::HGCalMulticluster *cl1, const l1t::HGCalMulticluster *cl2);
+
+bool compare_cluster_pt(const l1t::HGCalMulticluster *cl1, const l1t::HGCalMulticluster *cl2) {
```
put these functions in namespace?

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+  return std::make_pair(get_eta_bin(cl), get_phi_bin(cl));
+}
+
+class L1EGammaEEProducer : public edm::EDProducer {
```
legacy producers are not allowed

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+  const l1t::HGCalMulticlusterBxCollection &multiclusters = *multiclusters_h;
+
+  std::vector<const l1t::HGCalMulticluster *> selected_multiclusters;
+  std::map<std::pair<int, int>, std::vector<const l1t::HGCalMulticluster *>> etaphi_bins;
```
consider using unordered_map. hash for pair in this case could be a simple dense index (first + maxeta*second)

rekovic replied:
sure.  map works fine as well.

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+
+  // here we loop on the TPGs
+  for (auto cl3d = multiclusters.begin(0); cl3d != multiclusters.end(0); cl3d++) {
+    // std::cout << "-- CL3D is EG: " <<  cl3d->hwQual() << "   "<< cl3d->eta() <<"   "<< cl3d->pt()<<std::endl;
```
delete commented-out code

rekovic replied:
done.

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+    if (cl3d->hwQual()) {
+      if (cl3d->et() > minEt_) {
+        int hw_quality = 1;  // baseline EG ID passed
+        if (fabs(cl3d->eta()) >= 1.52) {
```
`std::abs()`, magic numbers

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  // std::cout << "# of selected HGCAL multiclusters: " << selected_multiclusters.size() << std::endl;
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+  // std::cout << "# of selected HGCAL multiclusters: " << selected_multiclusters.size() << std::endl;
+
+  std::sort(selected_multiclusters.begin(), selected_multiclusters.end(), compare_cluster_pt);
+  std::set<const l1t::HGCalMulticluster *> used_clusters;
```
consider using unordered_set

rekovic replied:
sure. set works fine as well.

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+
+  std::sort(selected_multiclusters.begin(), selected_multiclusters.end(), compare_cluster_pt);
+  std::set<const l1t::HGCalMulticluster *> used_clusters;
+  for (auto cl3d : selected_multiclusters) {
```
const auto&

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+  std::set<const l1t::HGCalMulticluster *> used_clusters;
+  for (auto cl3d : selected_multiclusters) {
+    if (used_clusters.find(cl3d) == used_clusters.end()) {
+      // reco::Candidate::LorentzVector mom = cl3d->p4();
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+      reco::Candidate::PolarLorentzVector mom_eint(
+          cl3d->iPt(l1t::HGCalMulticluster::EnergyInterpretation::EM), cl3d->eta(), cl3d->phi(), 0.);
+
+      // cl3d->pt()/(1+cl3d->hOverE())
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+      // this is not yet used
+      used_clusters.insert(cl3d);
+      auto eta_phi_bin = get_eta_phi_bin(cl3d);
+      // std::cout << " cl pt: " << cl3d->pt()
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+            // this bucket is not empty
+            for (auto other_cl_ptr : bucket->second) {
+              if (used_clusters.find(other_cl_ptr) == used_clusters.end()) {
+                if (fabs(other_cl_ptr->eta() - cl3d->eta()) < 0.02) {
```
std::abs

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+              if (used_clusters.find(other_cl_ptr) == used_clusters.end()) {
+                if (fabs(other_cl_ptr->eta() - cl3d->eta()) < 0.02) {
+                  if (fabs(reco::deltaPhi(other_cl_ptr->phi(), cl3d->phi())) < 0.1) {
+                    // std::cout << "MERGE with cl pt: " << other_cl_ptr->pt()
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+          auto bucket = etaphi_bins.find(std::make_pair(eta_bin, phi_bin));
+          if (bucket != etaphi_bins.end()) {
+            // this bucket is not empty
+            for (auto other_cl_ptr : bucket->second) {
```
const auto&?

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+      float calib_factor = calibrator_.calibrationFactor(mom.pt(), mom.eta());
+      l1t::EGamma eg =
+          l1t::EGamma(reco::Candidate::PolarLorentzVector(mom.pt() / calib_factor, mom.eta(), mom.phi(), 0.));
+      // l1t::EGamma eg(mom);
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+      l1t::EGamma eg =
+          l1t::EGamma(reco::Candidate::PolarLorentzVector(mom.pt() / calib_factor, mom.eta(), mom.phi(), 0.));
+      // l1t::EGamma eg(mom);
+      // FIXME: full duplication with HWqual 2
```
should it be fixed?

rekovic replied:
depricated. removed.

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+
+      l1t::EGamma eg_emint_brec =
+          l1t::EGamma(reco::Candidate::PolarLorentzVector(mom_eint.pt(), mom_eint.eta(), mom_eint.phi(), 0.));
+      // l1t::EGamma eg(mom);
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  iEvent.put(std::move(l1EgammaBxCollection), "L1EGammaCollectionBXVWithCuts");
```
is this tag needed since the producer doesn't produce anything else?

rekovic replied:
removed. done.

kpedro88 commented on L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py:
```diff
@@ -0,0 +1,185 @@
+    hfSF = cms.double(1.0),
+    debug = cms.bool(False),
+    skipCalibrations = cms.bool(False),
+    #debug = cms.bool(True),
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py:
```diff
@@ -0,0 +1,185 @@
+import FWCore.ParameterSet.Config as cms
+
+L1TowerCalibrationProducer = cms.EDProducer("L1TowerCalibrator",
```
fillDescriptions would be preferable

kpedro88 replied:
(for all producers introduced in this PR)

rekovic replied:
Propose to do it in the clean-up PR.

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+L1EGammaEECalibrator::L1EGammaEECalibrator(const edm::ParameterSet& pset) {
+  //read the JSON file and populate the eta - pt bins and the value container
+  boost::property_tree::ptree calibration_map;
+  read_json(pset.getParameter<edm::FileInPath>("calirationFile").fullPath(), calibration_map);
```
typo: calirationFile -> calibrationFile

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py:
```diff
@@ -0,0 +1,76 @@
+
+from Configuration.StandardSequences.Eras import eras
+
+process = cms.Process("L1AlgoTest",eras.Phase2_trigger)
```
please use the actual era for the phase 2 workflow being tested
(phase2_trigger is an internal use modifier, Phase2_trigger with that capitalization is not anything)

rekovic replied:
this is relic from previous versions of the integration branch

In 11_0_0_pre6  the l1 integration branch The driver config file which was used for testing uses the Phase2 era
https://github.com/cms-l1t-offline/cmssw/blob/phase2-l1t-integration-CMSSW_11_1_0_pre6/L1Trigger/L1TCommon/test/reprocess_L1TrackTrigger_L1T.py#L10

kpedro88 replied:
"please use the actual era for the phase 2 workflow being tested" does not mean "use the Phase2 era"

different scenarios have different associated eras, seen in https://github.com/cms-sw/cmssw/blob/f18f5a71e7278be55d457d17e38379b11cd04c19/Configuration/PyReleaseValidation/python/upgradeWorkflowComponents.py#L740

kpedro88 commented on L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py:
```diff
@@ -0,0 +1,76 @@
+process.GlobalTag = GlobalTag(process.GlobalTag, '100X_upgrade2023_realistic_v1', '')
+
+# Choose a 2023 geometry!
+process.load('Configuration.Geometry.GeometryExtended2023D17Reco_cff')
```
has this been tested with the D49 geometry used for the HLT TDR?

rekovic replied:
This file is an expert test configuration 

And yes, this code has been tested in the phase2 l1t integration branch with the recipe for D49 and D41 and has been used by the HLT people already to run on the HLT Phase2 samples.

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+#include "boost/property_tree/ptree.hpp"
+#include "boost/property_tree/json_parser.hpp"
+#include <iterator>
+// #include <iostream>
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+
+std::vector<float> as_vector(boost::property_tree::ptree const& pt, boost::property_tree::ptree::key_type const& key) {
+  std::vector<float> ret;
+  for (auto& item : pt.get_child(key))
```
const auto&?

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+  auto eta_h = as_vector(calibration_map, "eta_h");
+  eta_bins.insert(eta_h.back());
+
+  // std::cout << "# of eta bins: " << eta_bins.size()-1 << std::endl;
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+  std::copy(pt_l.begin(), pt_l.end(), std::inserter(pt_bins, pt_bins.end()));
+  auto pt_h = as_vector(calibration_map, "pt_h");
+  pt_bins.insert(pt_h.back());
+  // std::cout << "# of pt bins: " << pt_bins.size()-1 << std::endl;
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+#include <iterator>
+// #include <iostream>
+
+std::vector<float> as_vector(boost::property_tree::ptree const& pt, boost::property_tree::ptree::key_type const& key) {
```
put function in namespace?

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+
+int L1EGammaEECalibrator::get_bin(const std::set<float>& container, float value) const {
+  auto bin_l = container.upper_bound(value);
+  // std::cout << "value " << value << "lower boud " << *bin_l << " distance: " << std::distance(container.begin(), bin_l)-1 << std::endl;
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+    // value not mapped to any bin
+    return -1;
+  }
+  // return bin_l - container.begin();
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc:
```diff
@@ -0,0 +1,64 @@
+  if (bin_eta == -1 || bin_pt == -1)
+    return 1.;
+  auto n_bins_pt = pt_bins.size();
+  // std::cout << "pt: " << pt << " eta: " << eta << " ptbin: " << bin_pt << " etabin: " << bin_eta << " calib: " <<  calib_factors[(bin_eta*n_bins_pt)+bin_pt] << std::endl;
```
delete commented-out code

rekovic replied:
done

kpedro88 commented on L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc:
```diff
@@ -0,0 +1 @@
+#include "L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h"
```
any multi-line function implementations from the header file should be moved here

rekovic replied:
done

kpedro88 commented:
@rekovic a lot of these code review comments are the same ones that I make every time: no legacy producers, CMS code rules, basic cleanup, etc. It will speed the review and integration process for the remaining PRs if these issues are addressed before the PRs are submitted.

kpedro88 commented:
@rekovic what is the status of this PR?

kpedro88 commented:
@rekovic ping

rekovic commented:
Thanks @kpedro88 for the comments.
Addressing now.

rekovic committed:
Address ccoments with get_ and fabs.

rekovic committed:
Address comment delete commented out, use cmath and simplify return.

silviodonato commented:
@rekovic  just to avoid any misunderstanding, you have not yet pushed the commit into this PR. The last commit of this PR is of 27 days ago https://github.com/cms-sw/cmssw/pull/29904/commits !

silviodonato commented:
urgent

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16156

- This PR adds an extra 328KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16156/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16156/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
code-format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16161

- This PR adds an extra 328KB to repository

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f85da31156f5e6c4cf7a1f9e43343d1554919fb0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65df34/7094/summary.html
CMSSW: CMSSW_11_2_X_2020-06-15-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65df34/7094/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778757
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

rekovic commented:
@kpedro88 
Most of the comments are addressed. Still addressing remaining comments (replacing legacy producers) and will be hopefully done today/tomorrow, together with moving the data files.

rekovic committed:
Stream Producer and data file path, two producers in the L1 sequence..

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented:
test parameters:

 * pull_request = cms-data/L1Trigger-L1TCalorimeter#22

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16307

- This PR adds an extra 332KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30294


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16307/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16307/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
Code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16308

- This PR adds an extra 336KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30294

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: c2e1ba7b2aa9184f897ba18984e07d471b00e848

CMSSW: CMSSW_11_2_X_2020-06-21-0000
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-65df34/7229/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
20034.0 step2
<pre>runTheMatrix-results/20034.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D35+RecoFullGlobal_2026D35+HARVESTFullGlobal_2026D35.log

</pre>20434.0 step2
<pre>runTheMatrix-results/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D41+RecoFullGlobal_2026D41+HARVESTFullGlobal_2026D41.log

</pre>21234.0 step2
<pre>runTheMatrix-results/21234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D44_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D44+RecoFullGlobal_2026D44+HARVESTFullGlobal_2026D44.log

</pre>23234.0 step2
<pre>runTheMatrix-results/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D49_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D49+RecoFullGlobal_2026D49+HARVESTFullGlobal_2026D49/step2_TTbar_14TeV+TTbar_14TeV_TuneCP5_2026D49_GenSimHLBeamSpotFull14+DigiFullTrigger_2026D49+RecoFullGlobal_2026D49+HARVESTFullGlobal_2026D49.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

rekovic commented:
@mrodozov, @smuzaffar 

The above tests are failing due to missing data files which are available in my PR cms-data/L1Trigger-L1TCalorimeter#22
I guessed the following was enough to include the data PR into the testing of this PR
```
test parameters:

pull_request = cms-data/L1Trigger-L1TCalorimeter#22
```
but I guess not.  Can you please advise.  Thanks.

smuzaffar commented:
There us an extra ":" after test parameters

smuzaffar commented:
test parameters
 - pull_request = cms-data/L1Trigger-L1TCalorimeter#22

smuzaffar commented:
Please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/L1Trigger-L1TCalorimeter#22

cmsbuild commented:
+1
Tested at: c2e1ba7b2aa9184f897ba18984e07d471b00e848
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f559eb/7236/summary.html
CMSSW: CMSSW_11_2_X_2020-06-21-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

rekovic commented:
@smuzaffar 


> test parameters
> 
> * pull_request = [cms-data/L1Trigger-L1TCalorimeter#22](https://github.com/cms-data/L1Trigger-L1TCalorimeter/pull/22)

Thanks.

Can I have more than one PR in the list of test parameters ?  For example and additional PR which is not externals data ?

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f559eb/7236/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778760
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

kpedro88 commented:
@rekovic:
> Can I have more than one PR in the list of test parameters ? For example and additional PR which is not externals data ?

yes

rekovic commented on L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h:
```diff
@@ -0,0 +1,25 @@
+  float calibrationFactor(const float& pt, const float& eta) const;
+
+private:
+  int get_eta_bin(float eta) const { return get_bin(eta_bins, fabs(eta)); }
```
done both.

rekovic commented on L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:
```diff
@@ -0,0 +1,70 @@
+      etas.insert(etas.end(), etaBins.begin(), etaBins.end());
+      pts.insert(pts.end(), ptBins.begin(), ptBins.end());
+      scales.insert(scales.end(), scale.begin(), scale.end());
+      if (cpset.existsAs<std::vector<double>>("ptMin")) {
```
Why not use `existsAs`?  CMSSW is full of examples, over 1000 uses.

kpedro88 replied:
@Dr15Jones @makortel please comment on the use of `existsAs`

makortel replied:
Use of `existAs` is discouraged in favor of `fillDescriptions()`. With `existsAs` it becomes very difficult to figure out what exactly is the set of parameters for a module. Yes, it is unfortunately used in many places, but existing bad code is not a good excuse to add more bad code.

Dr15Jones replied:
I agree with Matti.

rekovic replied:
ok.  fair.  will implement.

rekovic replied:
done

rekovic replied:
done

rekovic commented on L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:
```diff
@@ -0,0 +1,70 @@
+        float ptMin = cpset.existsAs<double>("ptMin") ? cpset.getParameter<double>("ptMin") : 0;
+        ptMins = std::vector<float>(etaBins.size(), ptMin);
+      }
+      if (cpset.existsAs<std::vector<double>>("ptMax")) {
```
Why not use `existsAs`?  CMSSW is full of examples, over 1000 uses.

rekovic replied:
done

rekovic replied:
done

rekovic commented on L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:
```diff
@@ -0,0 +1,70 @@
+                             "Bad number of calibration scales, pts.size() * etas.size() != scales.size()");
+    }
+    float operator()(const float pt, const float abseta) const {
+      //std::cout << " ParametricCalibration : (pt=" << pt << ", abseta="<< abseta << ") : " << std::endl;
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:
```diff
@@ -0,0 +1,70 @@
+        }
+      }
+
+      //std::cout << "     ptBin = " << ptBin << std::endl;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+#include <memory>
+#include <array>
+#include <iostream>
+#include "TMath.h"
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+}
+
+bool L1EGCrystalClusterEmulatorProducer::passes_looseTkiso(float pt, float iso) {
+  if ((b0 + b1 * TMath::Exp(-b2 * pt) > iso))
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+static constexpr int n_crystals_towerPhi = 5;
+static constexpr int n_crystals_3towers = 3 * 5;
+static constexpr int n_towers_cardEta = 17;
+static constexpr int n_towers_Eta = 34;
```
Not that I know of.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+static constexpr float d0 = 0.96, d1 = 0.0003;                                         //passes_photon
+static constexpr float e0_looseTkss = 0.944, e1_looseTkss = 0.65, e2_looseTkss = 0.4;  //passes_looseTkss
+
+float getEta_fromL2LinkCardTowerCrystal(int link, int card, int tower, int crystal) {
```
This is not a getter of a class member.  It is a method that derives eta value from input information.  I am not aware that the CMS code [rules](http://cms-sw.github.io/cms_coding_rules.html#2----naming-rules-1) refer to that.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  return phimin;
+}
+
+class L1EGCrystalClusterEmulatorProducer : public edm::EDProducer {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h:
```diff
@@ -0,0 +1,70 @@
+      //std::cout << " ParametricCalibration : (pt=" << pt << ", abseta="<< abseta << ") : " << std::endl;
+      int ptBin = -1;
+      for (unsigned int i = 0, n = pts.size(); i < n; ++i) {
+        //std::cout << "     pt = " << pt << " , i = " << i << "  , pts[" << i << "] = " << pts[i] <<  std::endl;
```
done.

rekovic committed:
Address use M_PI.

rekovic committed:
Address remove non-used method.

rekovic committed:
Address magic numbers.

rekovic committed:
Address remove non-implemented method.

rekovic committed:
Address member variables end with trailing underscore.

rekovic committed:
Address remove non-used method.

rekovic committed:
Address simplifly and constexpr.

rekovic committed:
Address const auto&.

rekovic committed:
Address comment on separate line and magic number.

rekovic committed:
Remove commented code.

rekovic committed:
Address remove cout and use LogError and throw exception.

rekovic committed:
Code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+}
+
+bool L1EGCrystalClusterEmulatorProducer::passes_ss(float pt, float ss) {
+  if ((c0_ss + c1_ss * TMath::Exp(-c2_ss * pt)) > ss)
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+static constexpr int n_links_GCTcard = 48;
+static constexpr float ECAL_eta_range = 1.4841;
+static constexpr float half_crystal_size = 0.00873;
+static constexpr float PiValue = 3.14159;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+int get_insideEta(float eta, float centereta) {
+  float size_cell = 2 * ECAL_eta_range / (n_crystals_towerEta * n_towers_Eta);
+  if (eta - centereta < -1.0 * size_cell / 4)
+    return 0;
```
done. removed.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  if (id < n_towers_cardEta)
+    return id - n_towers_cardEta;
+  else
+    return id - 16;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  bool passes_he(float pt, float he);
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    int ctowerid;  // tower ID inside card (between 0 and 4*n_towers_cardEta-1)
+  };
+
+  bool order_clusters(mycluster c1, mycluster c2) { return c1.cpt < c2.cpt; }
```
done. removed.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+      return (position - other.position).mag();
+    };
+    bool operator==(SimpleCaloHit& other) const {
+      if (id == other.id && position == other.position && energy == other.energy && isEndcapHit == other.isEndcapHit)
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  iEvent.getByToken(ecalTPEBToken_, pcalohits);
+  std::vector<SimpleCaloHit> ecalhits;
+
+  for (auto& hit : *pcalohits.product()) {
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    if (hit.encodedEt() > 0)  // hit.encodedEt() returns an int corresponding to 2x the crystal Et
+    {
+      float et = hit.encodedEt() /
+                 8.;  // Et is 10 bit, by keeping the ADC saturation Et at 120 GeV it means that you have to divide by 8
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  edm::Handle<edm::SortedCollection<HcalTriggerPrimitiveDigi> > hbhecoll;
+  iEvent.getByToken(hcalTPToken_, hbhecoll);
+  for (auto& hit : *hbhecoll.product()) {
+    //if ( hit.SOI_compressedEt() == 0 ) continue;
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    if (et <= 0)
+      continue;
+    if (!(hcTopology_->validHT(hit.id()))) {
+      std::cout << " -- Hcal hit DetID not present in HCAL Geom: " << hit.id() << std::endl;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    hhit.id = hit.id();
+    hhit.id_hcal = hit.id();
+    hhit.position = hcal_tp_position;
+    //float et = hit.SOI_compressedEt() / 2.;
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    }
+    std::vector<HcalDetId> hcId = theTrigTowerGeometry.detIds(hit.id());
+    if (hcId.empty()) {
+      std::cout << "Cannot find any HCalDetId corresponding to " << hit.id() << std::endl;
```
done. replace with meessag logger and throw cms::Exception.

rekovic committed:
Address const auto&

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16363

- This PR adds an extra 368KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address comments in separate line, magic numbers, combine in one if statment.

rekovic committed:
Address magic numbers and Address comment on separate line.

rekovic committed:
Addressing put comments in separate lines. simplify max, magic numbers.

rekovic committed:
Code-format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+// shift all indices by 37 and loop over after 72
+int getToweriPhi_fromAbsoluteID(int id) {
+  if (id < n_towers_Phi / 2)
+    return id + 37;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+      if (isEndcapHit || other.isEndcapHit)
+        return 9999;  // We shouldn't compare integer indices in endcap, the map is not linear
+      // Logic from EBDetId::distancePhi() without the abs()
+      int PI = 180;
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    {
+      float et = hit.encodedEt() /
+                 8.;  // Et is 10 bit, by keeping the ADC saturation Et at 120 GeV it means that you have to divide by 8
+      if (et < 0.5)
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  std::vector<SimpleCaloHit> hcalhits;
+  edm::Handle<edm::SortedCollection<HcalTriggerPrimitiveDigi> > hbhecoll;
+  iEvent.getByToken(hcalTPToken_, hbhecoll);
+  for (auto& hit : *hbhecoll.product()) {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    if (hcId[0].subdetId() > 1)
+      continue;
+    GlobalVector hcal_tp_position = GlobalVector(0., 0., 0.);
+    for (auto& hcId_i : hcId) {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+  // Definition of L1 outputs
+  float ECAL_tower_L1Card[n_links_card][n_towers_cardEta]
+                         [n_towers_halfPhi];  // 36 L1 cards send each 4 links with 17 towers
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+  for (int cc = 0; cc < n_towers_halfPhi; ++cc) {  // Loop over 36 L1 cards
+
+    for (int nregion = 0; nregion < 6; ++nregion) {  // Loop over 3x4 etaxphi regions to search for max 5 clusters
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+              getCrystal_phiID(hit.position.phi()) >= getPhiMin_card(cc) &&
+              getCrystal_etaID(hit.position.eta()) <= getEtaMax_card(cc) &&
+              getCrystal_etaID(hit.position.eta()) >= getEtaMin_card(cc)) {  // Check that the hit is in the good card
+            if (getCrystal_etaID(hit.position.eta()) < getEtaMin_card(cc) + n_crystals_3towers * (nregion + 1) &&
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+          float n2x2_3 = 0;
+          float e2x2_4 = 0;
+          float n2x2_4 = 0;
+          for (auto& hit : ecalhits) {
```
not in this case

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+            }
+          }
+          mc1.c5x5 = e5x5;
+          mc1.c2x5 = e2x5_1;
```
done where applicable.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    // Merge clusters from different regions
+    for (unsigned int jj = 0; jj < unsigned(cluster_list[cc].size()); ++jj) {
+      for (unsigned int kk = jj + 1; kk < unsigned(cluster_list[cc].size()); ++kk) {
+        if (fabs(cluster_list[cc][jj].ceta - cluster_list[cc][kk].ceta) < 2 &&
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+        cluster_list[cc][jj].cpt =
+            cluster_list[cc][jj].cpt *
+            calib_(cluster_list[cc][jj].cpt,
+                   fabs(cluster_list[cc][jj].craweta));  //Mark's calibration as a function of eta and pt
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+          for (int ll = 0; ll < n_clusters_4link; ++ll) {  // We check the 12 clusters in the card on the right
+            if (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] < n_towers_cardEta &&
+                crystalID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] < 5 &&
+                fabs(5 * (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right]) % n_towers_cardEta +
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+  // Bremsstrahlung corrections. Merge clusters on the phi edges depending on pT (pt more than 10 percent, dphi leq 5, deta leq 1)
+  if (do_brem) {
+    for (int ii = 0; ii < 18; ++ii) {   // 18 borders in phi
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+        if (card_right > 35)
+          card_right = card_right - n_towers_halfPhi;
+        for (int kk = 0; kk < n_clusters_4link;
+             ++kk) {  // 12 clusters in the first card. We check the right side crystalID_cluster_L1Card[kk%4][kk/4][card_left]
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                       crystalID_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_left] % 5) <=
+                      1) {  //Distance of 1 max in eta
+                if (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] < n_towers_cardEta &&
+                    fabs(5 + crystalID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] / 5 -
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                       5 * (towerID_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_left]) %
+                           n_towers_cardEta -
+                       crystalID_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_left] % 5) <=
+                      1) {  //Distance of 1 max in eta
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+          energy_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_bottom] >
+              0) {                                       // If there is one cluster on the right side of the first card
+        for (int ll = 0; ll < n_clusters_4link; ++ll) {  // We check the card on the right
+          if (fabs(5 * (towerID_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_bottom] / n_towers_cardEta) +
```
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16364

- This PR adds an extra 372KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    }
+  }
+
+  vector<mycluster>
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+      int nclusters = 0;
+      bool build_cluster = true;
+
+      while (nclusters < 5 && build_cluster) {  // Continue until 5 clusters have been built or there is no cluster left
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+        if (build_cluster)
+          nclusters++;
+
+        if (build_cluster && nclusters > 0 && nclusters < 6) {  // Use only the 5 most energetic clusters
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                getCrystal_etaID(hit.position.eta()) >=
+                    getEtaMin_card(cc) +
+                        n_crystals_3towers *
+                            nregion) {  // clusters 3x5 in etaxphi using only the hits in the corresponding card and in the corresponding 3x4 region
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                  0) {  // if the tower is at the edge there might be brem corrections, whatever the crystal ID
+            for (int ll = 0; ll < n_clusters_4link; ++ll) {  // We check the 12 clusters in the card on the right
+              if (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] < n_towers_cardEta &&
+                  fabs(5 * (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right]) %
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                if (towerID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] < n_towers_cardEta &&
+                    fabs(5 + crystalID_cluster_L1Card[ll % n_links_card][ll / n_links_card][card_right] / 5 -
+                         (crystalID_cluster_L1Card[kk % n_links_card][kk / n_links_card][card_left] / 5)) <=
+                        5) {  //Distance of 5 max in phi
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+            }
+          }
+          if (do_brem && (rightlobe > 0.10 * mc1.cpt or leftlobe > 0.10 * mc1.cpt)) {
+            for (auto& hit : ecalhits) {
```
not in this case.  hit gets modfied 
`hit.used = true;`

rekovic committed:
Address comments on separate line.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  }
+
+  // Merge clusters on the eta edges
+  for (int ii = 0; ii < 18; ++ii) {  // 18 borders in eta
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+        }
+      }
+      for (int kk = 0; kk < n_towers_halfPhi; ++kk) {      // 36 cards
+        for (int ll = 0; ll < 4; ++ll) {                   // 4 links per card
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+            if (abs(etaOftower_fullDetector - cluster_etaOfTower_fullDetector) <= 2 &&
+                (abs(phiOftower_fullDetector - cluster_phiOfTower_fullDetector) <= 2 or
+                 abs(phiOftower_fullDetector - n_towers_Phi - cluster_phiOfTower_fullDetector) <=
+                     2)) {  // The towers are within 3. Needs to stitch the two phi sides together
```
done

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+            if (abs(etaOftower_fullDetector - cluster_etaOfTower_fullDetector) <= 2 &&
+                (abs(phiOftower_fullDetector - cluster_phiOfTower_fullDetector) <= 2 or
+                 abs(phiOftower_fullDetector - n_towers_Phi - cluster_phiOfTower_fullDetector) <=
+                     2)) {  // The towers are within 2. Needs to stitch the two phi sides together
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  const HcalTopology* hcTopology_;
+
+  struct mycluster {
+    float c2x2;
```
done.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  //First let's fill the towers
+  for (int ii = 0; ii < n_links_GCTcard; ++ii) {
+    for (int jj = 0; jj < n_towers_cardEta; ++jj) {
+      for (int ll = 0; ll < 3; ++ll) {
```
This is described by the comment on line 1021
`  //Reformat the information inside the 3 L2 cards`
`//First let's fill the towers`

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+  // Fill the cluster collection and towers as well
+  for (int ii = 0; ii < n_links_GCTcard; ++ii) {  // 48 links
+    for (int ll = 0; ll < 3; ++ll) {              // 3 cards
```
Again, the comment to the right describes that the loop is over 3 cards.  No need for constexpr here.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16377

- This PR adds an extra 72KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16377/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16377/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+          L1EGXtalClusters->push_back(cluster);
+
+          // BXVector l1t::EGamma quality defined with respect to these WPs
+          // FIXME, need to defaul some of these to 0 I think...
```
Deprecated comment. Removed.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  }
+
+  for (int ii = 0; ii < n_links_GCTcard; ++ii) {  // 48 links
+    for (int ll = 0; ll < 3; ++ll) {              // 3 cards
```
comment to the right clearly says looping over 3 cards.

kpedro88 replied:
And what if next year the plan changes to 4 cards? somebody has to go through and decide if every instance of the number 3 means 3 cards and should be changed? computers can automate this process by providing named constants that just have to be changed in one place.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+        // Add L1EGs if they match in iEta / iPhi
+        // L1EGs are already pT ordered, we will take the ID info for the leading one, but pT as the sum
+        for (auto l1eg : *L1EGXtalClusters) {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+          l1CaloTower.setNL1eg(n_L1eg++);
+          l1CaloTower.setL1egTowerEt(l1CaloTower.l1egTowerEt() + l1eg.pt());
+          if (l1CaloTower.nL1eg() > 1)
+            continue;  // Don't record L1EG quality info for subleading L1EG
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    if (!((a0_80 - a1_80 * pt) > iso))
+      return false;
+  }
+  if (pt >= 80) {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+}
+
+bool L1EGCrystalClusterEmulatorProducer::passes_photon(float pt, float pss) {
+  if (pss > d0 - d1 * pt)
```
done

rekovic committed:
Address remove deprecated comment, Address const auto&.

rekovic committed:
Address make_unique.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    }
+  }
+
+  std::unique_ptr<l1tp2::CaloCrystalClusterCollection> L1EGXtalClusters(new l1tp2::CaloCrystalClusterCollection);
```
done

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                                            is_looseTkiso&& is_looseTkss,
+                                            is_iso&& is_ss);
+          // Experimental parameters, don't want to bother with hardcoding them in data format
+          std::map<std::string, float> params;
```
I agree it could be used.  Map is fine too, no ?

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16378

- This PR adds an extra 376KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16378/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16378/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+int get_phi_bin(const l1t::HGCalMulticluster *cl) {
+  float phi_min = -M_PI;
+  float phi_max = M_PI;
+  unsigned n_phi_bins = 63;
```
done

rekovic committed:
Address constexpr.

rekovic committed:
Address remove commented code, code-format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic committed:
Address const auto&.

rekovic committed:
Address const auto&.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16379

- This PR adds an extra 376KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address remove commented code.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic committed:
Address const auto&.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16382

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address store pt, Address class private members, providing getters and setters.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+    bool isEndcapHit = false;  // If using endcap, we won't be using integer crystal indices
+
+    // tool functions
+    inline float pt() const { return (position.mag2() > 0) ? energy * sin(position.theta()) : 0.; };
```
done

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+
+  class SimpleCaloHit {
+  public:
+    EBDetId id;
```
done

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16384

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address constexpr.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc:
```diff
@@ -0,0 +1,190 @@
+}
+
+int get_eta_bin(const l1t::HGCalMulticluster *cl) {
+  float eta_min = 1.;
```
done

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16386

- This PR adds an extra 216KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address constexpr.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic committed:
Address remove commented code.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic committed:
Addres typo.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16394

- This PR adds an extra 376KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address use fillDescription and not existsAs.

rekovic committed:
Address remove non-needed forward declaration. code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16404

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

rekovic committed:
Address functions in namespace.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic committed:
Remove unneeded tag.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16405

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16405/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16405/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
Address move multiline functions.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16406

- This PR adds an extra 380KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16406/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16406/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic committed:
Address constexpr.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+        l1CaloTower.setTowerEta(getTowerEta_fromAbsoluteID(iEta_tower_L2Card[ii][jj][ll]));
+        l1CaloTower.setTowerPhi(getTowerPhi_fromAbsoluteID(iPhi_tower_L2Card[ii][jj][ll]));
+        // Some towers have incorrect eta/phi because that wasn't initialized in certain cases, fix these
+        if (l1CaloTower.towerEta() < -80. && l1CaloTower.towerPhi() < -90.) {
```
done

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+}
+
+bool L1EGCrystalClusterEmulatorProducer::passes_iso(float pt, float iso) {
+  if (pt < 80) {
```
done

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16407

- This PR adds an extra 392KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16407/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16407/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+                (abs(phiOftower_fullDetector - cluster_phiOfTower_fullDetector) <= 2 or
+                 abs(phiOftower_fullDetector - n_towers_Phi - cluster_phiOfTower_fullDetector) <=
+                     2)) {  // The towers are within 3. Needs to stitch the two phi sides together
+              if (!((cluster_phiOfTower_fullDetector == 0 && phiOftower_fullDetector == 71) or
```
too many. added comment.

rekovic committed:
Address magic numbers - add comment.

cmsbuild commented:
The code-checks are being triggered in jenkins.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+  //Second let's fill the clusters
+  for (int ii = 0; ii < n_towers_halfPhi; ++ii) {  // The cluster list is still in the L1 like geometry
+    for (unsigned int jj = 0; jj < unsigned(cluster_list_L2[ii].size()) && jj < n_clusters_4link; ++jj) {
+      crystalID_cluster_L2Card[n_links_card * (ii % n_clusters_4link) + jj % n_links_card][jj / n_links_card]
```
I rather would not touch these at this stage, till we have the automated comparison test in place.

rekovic commented on L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc:
```diff
@@ -0,0 +1,1216 @@
+      std::cout << " -- Hcal hit DetID not present in HCAL Geom: " << hit.id() << std::endl;
+      continue;
+    }
+    std::vector<HcalDetId> hcId = theTrigTowerGeometry.detIds(hit.id());
```
not sure what is suggested.

kpedro88 replied:
`const std::vector<HcalDetId>& hcId = theTrigTowerGeometry.detIds(hit.id());`

rekovic commented:
At this point, almost all the comments are addressed.   Will use this version to rebase onto https://github.com/cms-sw/cmssw/pull/30325

rekovic committed:
code format.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29904/16409

- This PR adds an extra 396KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  L1Trigger/Configuration/python/SimL1Emulator_cff.py  modified in PR(s): #30323, #30325
  - File  L1Trigger/L1CaloTrigger/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/L1EGammaEECalibrator.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/interface/ParametricCalibration.h  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/BuildFile.xml  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaCrystalsEmulatorProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/plugins/L1EGammaEEProducer.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1EGammaCrystalsEmulatorProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/L1TowerCalibrationProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaCrystalsProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/l1EGammaEEProducer_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/ntuple_cfi.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/python/test_L1EGCrystalClusterEmulator_cfg.py  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/L1EGammaEECalibrator.cc  modified in PR(s): #30325
  - File  L1Trigger/L1CaloTrigger/src/ParametricCalibration.cc  modified in PR(s): #30325

cmsbuild commented:
Pull request #29904 was updated. @cmsbuild, @rekovic, @benkrikler, @kpedro88 can you please check and sign again.

silviodonato commented:
moved to #30325
